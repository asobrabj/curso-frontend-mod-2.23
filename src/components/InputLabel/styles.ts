import styled from 'styled-components'

interface IInputLabelStyle {
  $fs?: 'xsmall' | 'small' | 'medium' | 'larger' | string
  $fStyle?: 'normal' | 'italic'
  $fw?: 'medium' | 'larger'
  $fFamily?: string
  $tAlign?: 'start' | 'end' | 'justify' | 'center'
  $tTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase'
  $color?: string
  $bgColor?: string
  $pd?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  $pdTop?: string
  $pdBottom?: string
  $pdRigth?: string
  $pdLeft?: string
  $mg?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  $mgTop?: string
  $mgBottom?: string
  $mgRigth?: string
  $mgLeft?: string
  $width?: string
  $height?: string
  $bdColor?: string
}

interface Defaults {
  [key: string]: string
}

const fontSize = (fs: string): string => {
  const defaults: Defaults = {
    xsmall: '14',
    small: '16',
    medium: '24',
    larger: '32',
  }
  const value = defaults[`${fs}`] || fs
  return `${value}px`
}

const fontWheigt = (fw: string): string => {
  const defaults: Defaults = {
    medium: '400',
    larger: 'bold',
  }
  const value = defaults[`${fw}`]
  return value
}

const paddingFn = (pd: string): string => {
  const defaults: Defaults = {
    xsmall: '5',
    small: '8',
    medium: '16',
    larger: '24',
    xlarger: '32',
  }
  const value = defaults[`${pd}`] ? `${defaults[`${pd}`]}px` : ''
  return value
}

const marginFn = (mg: string): string => {
  const defaults: Defaults = {
    xsmall: '5',
    small: '8',
    medium: '16',
    larger: '24',
    xlarger: '32',
  }
  const value = defaults[`${mg}`] ? `${defaults[`${mg}`]}px` : ''
  return value
}

export const Label = styled.label<IInputLabelStyle>`
  position: absolute;
  top: -10px;
  left: 10px;
  background-color: ${(props) => props.$bgColor || ''};

  span {
    color: red;
    margin-right: 3px;
  }
`

export const Input = styled.input<IInputLabelStyle>`
  padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '8px')};
  border: 1px solid #00abff;
  margin: ${(props) => (props.$mg ? marginFn(props.$mg) : '1px')};
  border-color: ${(props) => props.$bdColor || ''};
  width: 100%;
  background-color: '';

  &:focus {
    outline: none;
    border: 2px solid #00abff;
    margin: 0;
  }
`

export const InputLabel = styled.div<IInputLabelStyle>`
  font-size: ${(props) => (props.$fs ? fontSize(props.$fs) : '')};
  font-style: ${(props) => props.$fStyle || ''};
  font-weight: ${(props) => (props.$fw ? fontWheigt(props.$fw) : '')};
  font-family: ${(props) => props.$fFamily || ''};
  text-align: ${(props) => props.$tAlign || ''};
  text-transform: ${(props) => props.$tTransform || ''};
  color: ${(props) => props.$color || ''};
  padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '')};
  padding-top: ${(props) => (props.$pdTop ? props.$pdTop + 'px' : '')};
  padding-left: ${(props) => (props.$pdLeft ? props.$pdLeft + 'px' : '')};
  padding-right: ${(props) => (props.$pdRigth ? props.$pdRigth + 'px' : '')};
  padding-bottom: ${(props) => (props.$pdBottom ? props.$pdBottom + 'px' : '')};
  margin: ${(props) => (props.$mg ? marginFn(props.$mg) : '')};
  margin-top: ${(props) => (props.$mgTop ? props.$mgTop + 'px' : '')};
  margin-left: ${(props) => (props.$mgLeft ? props.$mgLeft + 'px' : '')};
  margin-right: ${(props) => (props.$mgRigth ? props.$mgRigth + 'px' : '')};
  margin-bottom: ${(props) => (props.$mgBottom ? props.$mgBottom + 'px' : '')};
  height: ${(props) => (props.$height ? props.$height + 'px' : '')};
  border-color: ${(props) => props.$bdColor || ''};
  position: relative;
  max-width: ${(props) => (props.$width ? props.$width + 'px' : '200px')};
  width: 100%;

  &.primary {
    ${Input} {
      border-radius: 8px;
    }
  }
  &.secondary {
    ${Input} {
      border-radius: 8px 0;
    }
  }

  &.error {
    color: ${(props) => props.$color || '#ff0000'};
    ${Input} {
      border: 1px solid #ff0000;
      &:focus {
        border: 2px solid #ff0000;
      }
    }
  }
`
