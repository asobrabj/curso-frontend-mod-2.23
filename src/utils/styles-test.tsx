import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
