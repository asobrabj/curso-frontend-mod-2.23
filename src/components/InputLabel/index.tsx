import * as S from './styles'

export interface IInputLabel {
  fs?: 'xsmall' | 'small' | 'medium' | 'larger' | string
  fStyle?: 'normal' | 'italic'
  fw?: 'medium' | 'larger'
  fFamily?: string
  tAlign?: 'start' | 'end' | 'justify' | 'center'
  tTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase'
  color?: string
  bgColor?: string
  pd?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  pdTop?: string
  pdBottom?: string
  pdRigth?: string
  pdLeft?: string
  mg?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  mgTop?: string
  mgBottom?: string
  mgRigth?: string
  mgLeft?: string
  className?: string
  width?: string
  height?: string
  placeholder?: string
  bdColor?: string
  type?:
    | 'text'
    | 'tel'
    | 'password'
    | 'email'
    | 'time'
    | 'date'
    | 'datetime-local'
    | 'radio'
    | 'checkbox'
    | 'file'
    | 'hidden'
    | 'number'
  required: boolean
  label: string
  id: string
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void
  onClick?: () => void
  value?: string
}

export default function InputLabel({
  fs,
  fStyle,
  fw,
  fFamily,
  tAlign,
  tTransform,
  color,
  bgColor,
  pd,
  pdTop,
  pdRigth,
  pdLeft,
  pdBottom,
  mg,
  mgTop,
  mgRigth,
  mgLeft,
  mgBottom,
  className,
  width,
  height,
  placeholder,
  bdColor,
  type = 'text',
  required = false,
  label,
  id,
  onChange,
  onClick,
  value,
}: IInputLabel) {
  return (
    <S.InputLabel
      className={`primary ${className}`}
      $fs={fs}
      $fw={fw}
      $fStyle={fStyle}
      $fFamily={fFamily}
      $tAlign={tAlign}
      $tTransform={tTransform}
      $color={color}
      $bgColor={bgColor}
      $pd={pd}
      $pdTop={pdTop}
      $pdRigth={pdRigth}
      $pdLeft={pdLeft}
      $pdBottom={pdBottom}
      $mg={mg}
      $mgTop={mgTop}
      $mgRigth={mgRigth}
      $mgLeft={mgLeft}
      $mgBottom={mgBottom}
      $width={width}
      $height={height}
      $bdColor={bdColor}
    >
      <S.Label htmlFor={`${id}-id`}>
        {required && <span>*</span>}
        {label}
      </S.Label>
      <S.Input
        title={`title-${id}`}
        type={type}
        placeholder={placeholder}
        id={`${id}-id`}
        required={required}
        name={id}
        onChange={onChange}
        onClick={onClick}
        value={value}
      />
    </S.InputLabel>
  )
}
