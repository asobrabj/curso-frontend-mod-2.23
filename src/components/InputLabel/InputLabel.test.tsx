import { fireEvent, screen } from '@testing-library/react'
import InputLabel from '.'
import { renderTheme } from '../../utils/styles-test'

describe('InputLabel Component', () => {
  it('deve renderizar corretamente', () => {
    const { container } = renderTheme(
      <InputLabel label="Nome" id="name" required />
    )
    expect(container).toMatchSnapshot()
  })

  it('deve chamar a função onChange ao digitar no input', () => {
    const onChangeMock = jest.fn()
    renderTheme(
      <InputLabel label="Nome" id="name" required onChange={onChangeMock} />
    )
    const input = screen.getByRole('textbox', {
      name: /Nome/i,
    }) as HTMLInputElement
    fireEvent.change(input, { target: { value: 'John' } })

    expect(input.value).toBe('John')
    expect(input.value).not.toBe('lucas')
  })

  it('deve chamar a função onClick ao clicar no input', () => {
    const onClickMock = jest.fn()
    renderTheme(
      <InputLabel label="Nome" id="name" required onClick={onClickMock} />
    )
    const input = screen.getByRole('textbox', {
      name: /Nome/i,
    }) as HTMLInputElement
    fireEvent.change(input, { target: { value: 'John' } })
    fireEvent.click(input)
    expect(onClickMock).toHaveBeenCalled()
  })

  it('deve exibir um asterisco (*) para campos obrigatórios', () => {
    const { getByText } = renderTheme(
      <InputLabel label="Nome" id="name" required />
    )
    const asterisk = getByText('*')
    expect(asterisk).toBeInTheDocument()
  })

  it('snapshot', () => {
    const onChangeMock = jest.fn()
    const { container } = renderTheme(
      <InputLabel
        label="Nome"
        id="name"
        required={false}
        onChange={onChangeMock}
      />
    )
    const input = screen.getByRole('textbox', {
      name: /Nome/i,
    }) as HTMLInputElement
    expect(input).not.toBeRequired()
    fireEvent.change(input, { target: { value: 'John' } })
    expect(container).toMatchSnapshot()
  })
})
