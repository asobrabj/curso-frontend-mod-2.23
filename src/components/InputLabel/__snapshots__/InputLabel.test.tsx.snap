// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputLabel Component deve renderizar corretamente 1`] = `
.c1 {
  position: absolute;
  top: -10px;
  left: 10px;
}

.c1 span {
  color: red;
  margin-right: 3px;
}

.c3 {
  padding: 8px;
  border: 1px solid #00abff;
  margin: 1px;
  width: 100%;
  background-color: '';
}

.c3:focus {
  outline: none;
  border: 2px solid #00abff;
  margin: 0;
}

.c0 {
  position: relative;
  max-width: 200px;
  width: 100%;
}

.c0.primary .c2 {
  border-radius: 8px;
}

.c0.secondary .c2 {
  border-radius: 8px 0;
}

.c0.error {
  color: #ff0000;
}

.c0.error .c2 {
  border: 1px solid #ff0000;
}

.c0.error .c2:focus {
  border: 2px solid #ff0000;
}

<div>
   
  <div
    class="c0 primary undefined"
  >
    <label
      class="c1"
      for="name-id"
    >
      <span>
        *
      </span>
      Nome
    </label>
    <input
      class="c2 c3"
      id="name-id"
      name="name"
      required=""
      title="title-name"
      type="text"
    />
  </div>
   
</div>
`;

exports[`InputLabel Component snapshot 1`] = `
.c1 {
  position: absolute;
  top: -10px;
  left: 10px;
}

.c1 span {
  color: red;
  margin-right: 3px;
}

.c3 {
  padding: 8px;
  border: 1px solid #00abff;
  margin: 1px;
  width: 100%;
  background-color: '';
}

.c3:focus {
  outline: none;
  border: 2px solid #00abff;
  margin: 0;
}

.c0 {
  position: relative;
  max-width: 200px;
  width: 100%;
}

.c0.primary .c2 {
  border-radius: 8px;
}

.c0.secondary .c2 {
  border-radius: 8px 0;
}

.c0.error {
  color: #ff0000;
}

.c0.error .c2 {
  border: 1px solid #ff0000;
}

.c0.error .c2:focus {
  border: 2px solid #ff0000;
}

<div>
   
  <div
    class="c0 primary undefined"
  >
    <label
      class="c1"
      for="name-id"
    >
      Nome
    </label>
    <input
      class="c2 c3"
      id="name-id"
      name="name"
      title="title-name"
      type="text"
    />
  </div>
   
</div>
`;
